### PR TITLE
Transactions API

### PR DESCRIPTION
**Pull request by mattsac**
Wrapping basic transactions.

### CODE REVIEW

File: src/muxy.ts [hunk 1]
```diff
@@ -197,6 +199,14 @@ export class Muxy implements MuxyInterface {
   public messenger: Messenger;
 
   /**
+   * Internal {@link Transactions}.
+   *
+   * @ignore
+   * @type {Transactions}
+   */
+   public transaction: PurchaseClient;
```
Comment (camr) [inline src/muxy.ts:207]: This field shouldn't be called `transaction` anymore. `purchaseClient` works for me.

File: src/muxy.ts [hunk 2]
```diff
@@ -520,6 +530,10 @@ export class Muxy implements MuxyInterface {
       Muxy.printInfo();
     }
 
+    if (options.transactionsEnabled) {
+      this.transaction.onUserPurchase(() => {});
```
Comment (camr) [inline src/muxy.ts:534]: This callback should send the message to Muxy's backend. And that call should be removed from the default implementation.

File: src/purchase-client.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
+import { MuxySDK } from ".";
+import { CurrentEnvironment, forceType } from './util';
+import Config from './config';
+
+import { TwitchBitsProduct, TwitchBitsTransaction } from './twitch';
+
+export enum TransactionType {
```
Comment (camr) [inline src/purchase-client.ts:7]: `PurchaseClientType`

File: src/purchase-client.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+export interface Product {
+  sku: string;
+  displayName: string;
+  cost: Cost;
+}
+
+// TwitchTransactions implements the basic 'transaction' interface.
+// This is used by SDK to provide low-level access to twitch bits transactions.
+class TwitchPurchase implements PurchaseClient {
```
Comment (camr) [inline src/purchase-client.ts:34]: `TwitchPurchaseClient` and comment needs to be updated.

File: src/purchase-client.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+  Server,
+  Unknown
+}
+
+export interface PurchaseClient {
+  sdk: MuxySDK;
+
+  getProducts(): Promise<Product[]>;
+  purchase(identifier: string): void;
+  onUserPurchase(callback: <T>(data: T) => void): void;
```
Comment (camr) [inline src/purchase-client.ts:18]: Rather than `callback` taking a generic argument, you should define a base `Transaction` interface (that conforms to the `TwitchBitsTransaction` type) and have that passed as the parameter to the callback.

File: src/purchase-client.ts [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+   * @since 2.4.0
+   *
+   * @param callback a function(body)
+   *
+   * @example
+   * client.onUserPurchase(() => {
+   *  console.log("Transaction finished!");
+   * });
+   */
+  public onUserPurchase(callback): void {
```
Comment (camr) [inline src/purchase-client.ts:93]: `callback` needs a type. See my comment above about expected parameter.

File: src/purchase-client.ts [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    window.Twitch.ext.bits.onTransactionComplete(
+      (tx: TwitchBitsTransaction) => {
+        if (tx.initiator.toLowerCase() === "current_user") {
+          this.sdk.signedRequest("POST", "bits/transactions", JSON.stringify(tx));
```
Comment (camr) [inline src/purchase-client.ts:97]: This line should only be done once, so set it up in the default handler implementation when `transactionsEnabled` i true.

**Review by camr — CHANGES_REQUESTED** [review]

File: src/muxy.ts (hunk 2 above)
Comment (mattsac) [reply src/muxy.ts:534]: So where is this getting the tx from (that's used in that call in the default implementation)

Comment (camr) [reply src/muxy.ts:534]:
```this.transaction.onUserPurchase((tx: GenericTransaction) => {
  // sdk.dothething()
})```

File: src/muxy.ts [hunk 8]
```diff
@@ -30,6 +31,7 @@ export interface SDKMap {
  */
 export interface SetupOptions {
   clientID: string;
+  purchaseClientEnabled: boolean;
```
Comment (camr) [inline src/muxy.ts:34]: Don't call this `purchaseClientEnabled`, that is an internal name not one the user needs to be concerned with. Please use `transactionsEnabled` as before.

File: src/muxy.ts [hunk 9]
```diff
@@ -197,6 +199,22 @@ export class Muxy implements MuxyInterface {
   public messenger: Messenger;
 
   /**
+   * Internal {@link purchaseClient}.
```
Comment (camr) [inline src/muxy.ts:202]: This `@link` would link to itself. Notice how `messenger` links to the `Messenger` class.

File: src/purchase-client.ts [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+  transactionId: string;
+  product: Product;
+  userId: string;
+  displayName: string;
+  initiator: any;
+  transactionReceipt: string;
+}
+
+// TwitchTransactions implements the basic 'transaction' interface.
+// This is used by SDK to provide low-level access to twitch bits transactions.
```
Comment (camr) [inline src/purchase-client.ts:42]: Comment needs to be updated.

File: src/purchase-client.ts (hunk 7 above)
Comment (camr) [reply src/purchase-client.ts:106]: This still needs to be removed from here.

File: src/sdk.ts [hunk 11]
```diff
@@ -393,15 +395,24 @@ export default class SDK {
...
-        mxy.debugOptions
+        mxy.debugOptions,
       );
 
       mxy.SDKClients[identifier] = this;
     }
 
+    if (mxy.purchaseClientEnabled) {
+      this.purchaseClient.onUserPurchase((tx: Transaction) => {
+        if (tx.initiator.toLowerCase() === "current_user") {
```
Comment (camr) [inline src/sdk.ts:410]: This check doesn't need to be done here, as it is already performed inside `onUserPurchase`.

File: src/sdk.ts [hunk 12]
```diff
@@ -1352,6 +1363,75 @@ export default class SDK {
...
+   * 
+   * @example
+   * const productList = sdk.getProducts();
+   */
+   public products() {
+    if (!mxy.didLoad) {
+      throw new Error('sdk.loaded() was not complete. Please call this method only after the promise has resolved.');
+    }
+
+    return this.transaction.getProducts();
```
Comment (camr) [inline src/sdk.ts:1388]: `this.transaction` doesn't exist anymore.

Comment (camr) [inline src/sdk.ts:1388]: Since it returns a Promise, this function needs to be listed as `async` and its return type set and documented correctly.

File: src/sdk.ts [hunk 13]
```diff
@@ -1667,6 +1748,14 @@ export default class SDK {
     this.messenger = messenger;
 
     /**
+     * The backend transaction client.
+     * @private
+     * @type {Transaction}
+     *
+     */
+    this.transaction = transaction;
```
Comment (camr) [inline src/sdk.ts:1756]: This shouldn't be called transaction.

File: src/purchase-client.ts (hunk 10 above)
Comment (camr) [reply src/purchase-client.ts:42]: This still mentions that it implements the `transaction` interface.

File: src/purchase-client.ts [hunk 14]
```diff
@@ -0,0 +1,203 @@
+import { MuxySDK } from ".";
+import { CurrentEnvironment } from './util';
+import Config from './config';
+
+import { TwitchBitsProduct, TwitchBitsTransaction } from './twitch';
+
+// Muxy dev variable
+declare global {
+  interface Window {
+    MEDKIT_PURCHASABLE_ITEMS: Product[] | PromiseLike<Product[]>;
```
Comment (camr) [inline src/purchase-client.ts:10]: I don't think this should ever by a Promise. What's the reasoning here?

File: src/purchase-client.ts [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+   * @example
+   * const products = client.getProducts();
+   */
+  public getProducts(): Promise<TwitchBitsProduct[]> {
+    const target = window.Twitch?.ext?.bits;
+    if (!window.Twitch.ext.bits) {
+      throw new Error('Twitch helper was not loaded.');
+    }
+    
+    return target.getProducts();
```
Comment (camr) [inline src/purchase-client.ts:90]: You probably don't need the assignment to `target` unless you also want to change your check to `if (!target)`

File: src/purchase-client.ts [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+   *
+   * @param callback a function(body)
+   *
+   * @example
+   * client.onUserPurchase(() => {
+   *  console.log("Transaction finished!");
+   * });
+   */
+  public onUserPurchase(callback: (tx: Transaction) => void): void {
+    window.Twitch.ext.bits.onTransactionComplete(
```
Comment (camr) [inline src/purchase-client.ts:106]: In `getProducts` you check if `window.Twitch.ext.bits` exists. Why not here?

File: src/purchase-client.ts [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+   *
+   * @throws {TwitchHelperError} Will throw an error if the Twitch Helper didn't load.
+   * 
+   * @param identifier the twitch sku of the item to be purchased.
+   *
+   * @example
+   * client.purchase("XXSKU000");
+   */
+  public purchase(identifier: string) {
+    if (!window.Twitch.ext.bits) {
```
Comment (camr) [inline src/purchase-client.ts:63]: This check should have the optional chaining (`?.`) so as to not throw an error if one along the path is undefined.

File: src/purchase-client.ts [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+      throw new Error('Twitch helper was not loaded.');
+    }
+
+    window.Twitch?.ext?.bits.useBits(identifier);
```
Comment (camr) [inline src/purchase-client.ts:67]: You don't need the optional chaining here if you do the check above.

File: src/purchase-client.ts [hunk 19]
```diff
@@ -0,0 +1,203 @@
...
+          callback(tx);
+        }
+      }
+    );
+  }
+}
+
+// DevPurchaseClient implements the basic 'purchase client' interface.
+// This is used by SDK to provide low-level access to stubbed transactions.
+class DevPurchaseClient implements PurchaseClient {
```
Comment (camr) [inline src/purchase-client.ts:118]: You'll need a field of this class to track stored callbacks.

File: src/purchase-client.ts [hunk 20]
```diff
@@ -0,0 +1,203 @@
...
+   *
+   * @param callback a function(body)
+   *
+   * @example
+   * client.onUserPurchase(() => {
+   *  console.log("Transaction finished!");
+   * });
+   */
+  public onUserPurchase(callback: (tx: Transaction) => void): void {
+    callback;
```
Comment (camr) [inline src/purchase-client.ts:189]: This should store the supplied callback in a list on this class.

File: src/purchase-client.ts [hunk 21]
```diff
@@ -0,0 +1,203 @@
...
+   *   }
+   * ];
+   *
+   * const products = client.getProducts();
+   */
+  public getProducts(): Promise<Product[]> {
+    if ("MEDKIT_PURCHASABLE_ITEMS" in window) {
+      return new Promise((resolve, reject) => {
+        resolve(window.MEDKIT_PURCHASABLE_ITEMS);
+      });
```
Comment (camr) [inline src/purchase-client.ts:168]: You can shorten this to `return Promise.resolve(window.MEDKIT_PURCHASABLE_ITEMS)`

File: src/purchase-client.ts [hunk 22]
```diff
@@ -0,0 +1,203 @@
...
+    }
+
+    return new Promise((resolve, reject) => {
+      resolve([]);
+    });
```
Comment (camr) [inline src/purchase-client.ts:173]: Same shortening to `return Promise.resolve([])`

File: src/purchase-client.ts [hunk 23]
```diff
@@ -0,0 +1,203 @@
...
+   *
+   * @throws {SDKError} Will throw an error if the MuxySDK didn't load.
+   * 
+   * @param identifier the twitch sku of the item to be purchased.
+   *
+   * @example
+   * client.purchase("TESTSKU00");
+   */
+  public purchase(identifier: string) {
+    // TODO: implement some type of purchase for dev testing?
```
Comment (camr) [inline src/purchase-client.ts:132]: Loop through all stored callbacks and call each one providing a synthesized `Transaction`.

Comment (camr) [reply src/purchase-client.ts:132]: The synthesized transaction should look through `window.MEDKIT_PURCHASABLE_ITEMS` for the proper sku and pull info from that.

File: src/purchase-client.ts [hunk 24]
```diff
@@ -0,0 +1,203 @@
...
+   * @since 2.4.0
+   *
+   * @throws {SDKError} Will throw an error if the MuxySDK didn't load.
+   * 
+   * @param identifier the twitch sku of the item to be purchased.
+   *
+   * @example
+   * client.purchase("TESTSKU00");
+   */
+  public purchase(identifier: string) {
```
Comment (camr) [inline src/purchase-client.ts:131]: `identifier` should be `sku`

File: src/purchase-client.ts [hunk 25]
```diff
@@ -0,0 +1,203 @@
...
+   * Returns a list of Twitch Bits Products for the current channel.
+   * 
+   * @async
+   * @since 2.4.0
+   *
+   * @return {Promise<[]Product>} Resolves with an array of {@link Product}
+   * objects for each available sku.
+   *
+   * @example
+   * window.MEDKIT_PURCHASE_ITEMS = [{
```
Comment (camr) [inline src/purchase-client.ts:145]: Field is `window.MEDKIT_PURCHASABLE_ITEMS`

**Review by camr — CHANGES_REQUESTED** [review]

File: src/purchase-client.ts (hunk 14 above)
Comment (mattsac) [reply src/purchase-client.ts:10]: Doesn't have to.  Was making it work the same way as the twitch version.

File: src/sdk.ts [hunk 26]
```diff
@@ -1352,6 +1362,76 @@ export default class SDK {
...
+   * @async
+   * @since 2.4.0
+   *
+   * @throws {SDKError} Will throw an error if the MuxySDK didn't load.
+   * 
+   * @return {Promise<[]Product>} Resolves with an array of {@link Product}
+   * objects for each available sku.
+   * 
+   * @example
+   * const productList = sdk.getProducts();
```
Comment (camr) [inline src/sdk.ts:1381]: Function is galled `products` (should it be `getProducts`?), and returns a Promise. Should be: `const products = await sdk.products()`

File: src/state-client.ts [hunk 27]
```diff
@@ -491,16 +491,14 @@ class StateClient {
     this.signedRequest<undefined, TriviaTeam>(identifier, 'GET', 'team_membership');
 
   /** @ignore */
-  public addExtensionTriviaQuestion = (identifier: string, triviaQuestion: TriviaQuestion) =>  {
+  public addExtensionTriviaQuestion = (identifier: string, triviaQuestion: TriviaQuestion) =>  
     // TODO: determine actual response type and replace unknown.
```
Comment (camr) [inline src/state-client.ts:495]: You'll need to remove the comment for this to work.

File: src/state-client.ts [hunk 28]
```diff
@@ -491,16 +491,14 @@ class StateClient {
...
     this.signedRequest<TriviaQuestion, unknown>(identifier, 'POST', 'curated_poll_edit', triviaQuestion);
```
Comment (camr) [inline src/state-client.ts:496]: This `unknown` can now be typed correctly.

File: src/state-client.ts [hunk 29]
```diff
@@ -491,16 +491,14 @@ class StateClient {
...
-  }
 
   /** @ignore */
-  public removeExtensionTriviaQuestion = (identifier: string, triviaQuestionID: string) => {
+  public removeExtensionTriviaQuestion = (identifier: string, triviaQuestionID: string) => 
     // TODO: determine actual response type.
```
Comment (camr) [inline src/state-client.ts:500]: You'll need to remove the comment for this to work.

File: src/state-client.ts [hunk 30]
```diff
@@ -491,16 +491,14 @@ class StateClient {
...
     this.signedRequest<{ id: string }, unknown>(identifier, 'DELETE', 'curated_poll_edit', { id: triviaQuestionID });
```
Comment (camr) [inline src/state-client.ts:501]: This `unknown` can now be typed correctly.

File: src/state-client.ts [hunk 31]
```diff
@@ -512,15 +510,14 @@ class StateClient {
     );
 
   /** @ignore */
-  public removeExtensionTriviaOptionFromQuestion = (identifier: string, questionID: string, optionID: string) => {
+  public removeExtensionTriviaOptionFromQuestion = (identifier: string, questionID: string, optionID: string) => 
     // TODO: determine actual response type and replace unknown.
```
Comment (camr) [inline src/state-client.ts:514]: Remove the comment.

File: src/state-client.ts [hunk 32]
```diff
@@ -512,15 +510,14 @@ class StateClient {
...
     this.signedRequest<{ question: string, option: string }, unknown>(
```
Comment (camr) [inline src/state-client.ts:515]: Type `unknown`.

File: src/state-client.ts [hunk 33]
```diff
@@ -537,10 +534,9 @@ class StateClient {
     );
 
   /** @ignore */
-  public setExtensionTriviaQuestionVote = (identifier: string, questionID: string, optionID: string) => {
+  public setExtensionTriviaQuestionVote = (identifier: string, questionID: string, optionID: string) => 
     // TODO: determine actual response type and replace unknown.
```
Comment (camr) [inline src/state-client.ts:538]: Remove comment.

File: src/state-client.ts [hunk 34]
```diff
@@ -537,10 +534,9 @@ class StateClient {
...
     this.signedRequest<{ question_id: string, vote: string }, unknown>(identifier, 'POST', 'curated_poll', { question_id: questionID, vote: optionID });
```
Comment (camr) [inline src/state-client.ts:539]: Type `unknown`.

File: src/sdk.ts [hunk 35]
```diff
@@ -1352,6 +1362,76 @@ export default class SDK {
...
+      throw new Error('sdk.loaded() was not complete. Please call this method only after the promise has resolved.');
+    }
+
+    return this.purchaseClient.getProducts();
+  }
+
+  /**
+   * Starts transaction for a specific product identifier.
+   *
+   * @since 2.4.0
```
Comment (camr) [inline src/sdk.ts:1394]: This function is `@async`

File: __tests__/purchase-client.test.ts [hunk 36]
```diff
@@ -0,0 +1,49 @@
...
+        {
+          sku: "UNITTESTSKU02",
+          displayName: "Unit Test SKU 02",
+          cost: {
+           amount: 2,
+           type: "test-cost",
+          }
+        }
+      ];
+    });
```
Comment (camr) [inline __tests__/purchase-client.test.ts:23]: I thought we discussed moving this to a constant outside of the functions?

File: __tests__/sdk.test.ts [hunk 37]
```diff
@@ -28,6 +29,7 @@ describe('SDK', () => {
     client = new StateClient(prom, new DefaultDebugOptions());
     messenger = Messenger(new DefaultDebugOptions());
     messenger.close();
+    purchaseClient = PurchaseClient();
```
[the remark this answers is not in the document]
Comment (mattsac) [reply __tests__/sdk.test.ts:32]: I see no errors, and the test that would be using it passes, so...

File: __tests__/purchase-client.test.ts (hunk 36 above)
Comment (mattsac) [reply __tests__/purchase-client.test.ts:23]: Yes...a while after I checked this in.

Comment (camr) [reply __tests__/purchase-client.test.ts:23]: So are you still working on this?

File: __tests__/sdk.test.ts (hunk 37 above)
[the remark this answers is not in the document]
Comment (camr) [reply __tests__/sdk.test.ts:32]: I mean. `PurchaseClient` is an interface. You can't call an interface. So...

[the remark this answers is not in the document]
Comment (mattsac) [reply __tests__/sdk.test.ts:32]: I get that.  I'm saying yes, it seems to compile.

File: src/purchase-client.ts [hunk 38]
```diff
@@ -0,0 +1,213 @@
...
+    if ("MEDKIT_PURCHASABLE_ITEMS" in window) {
+      const devItems = this.getProducts();
+
+      const products = Object.keys(devItems).map((sku) => devItems[sku]);
+      const foundItem = products.find((sku: string) => sku === sku);
+
+      setTimeout(() => {
+        if (foundItem) {
+          this.purchaseCallbacks.forEach(function(callback){
+            callback(foundItem);
```
[the remark this answers is not in the document]
Comment (mattsac) [reply src/purchase-client.ts:142]: Do we care what the other values are for the Transaction?  (just so they look like valid, from the server data)

```
Transaction = {
  transactionId: string;
  product: Product;
  userId: string;
  displayName: string;
  initiator: any;
  transactionReceipt: string;
}
```

File: src/state-client.ts [hunk 39]
```diff
@@ -218,6 +219,10 @@ class StateClient {
     this.token = token;
   }
 
+  public sendTransactionToServer(identifier: string, tx: Transaction) {
+    this.signedRequest<Transaction, Record<string, unknown>>(identifier, "POST", "bits/transaction", tx);
+  }
+
```
Comment (camr) [inline src/state-client.ts:225]: Can you move this down to the bottom of this file (with the other simple wrappers) and add an `@ignore` comment like the others?

Comment (camr) [reply src/state-client.ts:225]: The endpoint is actually `bits/transactions` (plural). I think I typed that wrong in Slack.

File: __tests__/sdk.test.ts (hunk 37 above)
[the remark this answers is not in the document]
Comment (camr) [reply __tests__/sdk.test.ts:32]: Still need to change this to `DefaultClient` and/or create a `TestingPurchaseClient` directly.

File: src/config.ts [hunk 40]
```diff
@@ -49,6 +50,21 @@ export default class Config {
...
+      case ENVIRONMENTS.ADMIN: // Currently unable to hook into the twitch pubsub system from admin
+      case ENVIRONMENTS.SANDBOX_ADMIN:
+      case ENVIRONMENTS.SANDBOX_TWITCH:
+      case ENVIRONMENTS.PRODUCTION:
+        return PurchaseClientType.Twitch;
+      case ENVIRONMENTS.SERVER:
+        return PurchaseClientType.Server;
+    }
+    return PurchaseClientType.Unknown;
+  }
```
Comment (camr) [inline src/config.ts:66]: The `ENVIRONMENTS` things is deprecated (you can read about it in the corresponding file), but I believe there is one for when running in the unit test environment. Since you have a testing-specific type and class, you should see about adding it here.

File: src/purchase-client.ts [hunk 41]
```diff
@@ -0,0 +1,322 @@
...
+   * client.onUserPurchase((transaction) => {
+   *  console.log("Transaction finished!");
+   * });
+   */
+  public onUserPurchase(callback: (tx: Transaction) => void): void {
+    this.purchaseCallbacks.push(callback);
+  }
+}
+export default function DefaultPurchaseClient(): PurchaseClient {
+  const type = Config.DefaultPurchaseClientType(CurrentEnvironment());
```
Comment (camr) [inline src/purchase-client.ts:311]: Can `Config.DefaultPurchaseClientType` ever return `PurchaseClientType.Test`?

**Review by camr — CHANGES_REQUESTED** [review]

File: __tests__/sdk.test.ts [hunk 42]
```diff
@@ -28,6 +30,7 @@ describe('SDK', () => {
     client = new StateClient(prom, new DefaultDebugOptions());
     messenger = Messenger(new DefaultDebugOptions());
     messenger.close();
+    purchaseClient = TestPurchaseClient();
```
Comment (camr) [inline __tests__/sdk.test.ts:33]: Need a `new` before this call.

File: src/purchase-client.ts [hunk 43]
```diff
@@ -0,0 +1,322 @@
...
+        }
+      }
+    );
+  }
+}
+
+// DevPurchaseClient implements the basic 'purchase client' interface.
+// This is used by SDK to provide low-level access to stubbed transactions.
+class DevPurchaseClient implements PurchaseClient {
+  private purchaseCallbacks: Array<Function> = [];
```
Comment (camr) [inline src/purchase-client.ts:124]: Add a more specific type for callbacks: `Array<(tx: Transaction) => void>`

File: src/state-client.ts [hunk 44]
```diff
@@ -552,6 +545,11 @@ class StateClient {
   /** @ignore */
   public getExtensionTriviaLeaderboard = (identifer: string) =>
     this.signedRequest<undefined, TriviaLeaderboard>(identifer, 'GET', 'curated_poll_leaderboard');
+
+  /** @ignore */
+  public sendTransactionToServer(identifier: string, tx: Transaction) {
+    this.signedRequest<Transaction, Record<string, unknown>>(identifier, "POST", "bits/transactions", tx);
+  }
```
Comment (camr) [inline src/state-client.ts:552]: With the braces, this won't automatically return the result of `signedRequest`. You can either remove them and add a `=>` as above, or add an explicit `return` statement before the call to `this.signedRequest`.

**Review by camr — CHANGES_REQUESTED** [review]

File: __tests__/sdk.test.ts [hunk 45]
```diff
@@ -1,17 +1,19 @@
 import Analytics from '../src/analytics';
 import { DebugOptions, DefaultDebugOptions } from '../src/debug';
 import Messenger from '../src/messenger';
+import PurchaseClient from '../src/purchase-client';
```
Comment (camr) [inline __tests__/sdk.test.ts:4]: Default import from this file, should use `{ PurchaseClient }`

File: src/config.ts [hunk 46]
```diff
@@ -36,55 +37,74 @@ export default class Config {
...
         return AuthorizationFlowType.TestAuth;
-
-      case ENVIRONMENTS.SANDBOX_ADMIN:
-      case ENVIRONMENTS.ADMIN:
+      case Util.Environments.Admin :
         return AuthorizationFlowType.AdminAuth;
-
-      case ENVIRONMENTS.SANDBOX_TWITCH:
-      case ENVIRONMENTS.PRODUCTION:
+      case Util.Environments.SandboxAdmin:
```
Comment (camr) [inline src/config.ts:78]: This looks like you may have moved sandbox admin to the wrong spot. It should end up returning `AdminAuth`

**Review by camr — CHANGES_REQUESTED** [review]

**Review by camr — APPROVED** [review]

**Review by albert-wang — APPROVED** [review]

File: src/purchase-client.ts [hunk 47]
```diff
@@ -0,0 +1,322 @@
...
+      const foundItem = products.find((sku: string) => sku === sku);
+
+      setTimeout(() => {
+        if (foundItem) {
+          const tx: Transaction = {
+            transactionId: "dev-transaction-id",
+            product: foundItem,
+            userId: "dev-test-user",
+            displayName: "DevTestUser",
+            initiator: "CURRENT_USER",
```
[the remark this answers is not in the document]
Comment (camr) [reply src/purchase-client.ts]: The Twitch docs do say that it will always be uppercase. From https://dev.twitch.tv/docs/extensions/reference
```
Valid values:

    CURRENT_USER (the user who executed, or initiated, the transaction).
    OTHER (a user other than CURRENT_USER).
```

But as long as we're lower/uppercasing it and checking that, I guess it doesn't matter.
